### ISSUE TITLE
Is it possible to set the batch size of the calibration_data_loader indepentent from the batch size of the exported model?

### ISSUE THREAD

**Issue by kfrohwergmh** [ultralytics/ultralytics#24470]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I'm wondering if it's possible to adjust the batch size for the calibration dataset's data loader independent of batch size of the exported model. 
Nvidia itself recommends keeping the batch size as large as possible for int8 quantization (preferably in a single batch, if the GPU allows it) to make the quantization more accurate.

Source: https://forums.developer.nvidia.com/t/what-batch-size-to-use-for-post-training-quantization-int8-calibration/260014

My understanding of the existing `batch` parameter for export is that it affects the batch dimension of the exported model and the batch size of the calibration data set, if provided. 
Is that correct? That would mean that my model’s output increases from [1, 4+classes, 8400] to [n, 4+classes, 8400], where n is the size I define as the batch.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @kfrohwergmh, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your question routed quickly, and an Ultralytics engineer will also assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training or export ❓ Question, please provide as much information as possible, including your exact export command or code snippet, model type, target format, device, logs, and any calibration dataset details so the team can better understand the behavior you're seeing 🔍.

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by raimbekovm** [comment]
@kfrohwergmh your understanding is correct. The `batch` argument is currently shared:

- It sets the exported model's batch dim (the dummy input is `torch.zeros(batch, C, H, W)` - see [`exporter.py#L435`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/exporter.py#L435)), so the output becomes `[batch, 4+classes, 8400]` as you described.
- It also sets the calibration dataloader's batch size (see [`get_int8_calibration_dataloader`](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/exporter.py#L564-L591)), which the TensorRT INT8 calibrator picks up via `dataset.batch_size`.

There is no separate knob today to decouple them.

The workaround that gets you close to NVIDIA's recommendation is `dynamic=True` combined with a larger `batch`:

```python
model.export(format='engine', int8=True, dynamic=True, batch=16, data='coco.yaml')
```

This builds an optimization profile of `min=(1, C, 32, 32)` ... `opt=(batch, C, H, W)`, so the engine accepts any batch in `[1, batch]` at inference time, while calibration runs with the larger batch you specified, which is what improves INT8 scale estimation. The caveat is that your runtime max batch is capped at the value you exported with.

If a strictly fixed batch (e.g. `batch=1`) for the deployed engine while calibrating with a larger batch is important for your use case, that would need a new export argument (e.g. `calibration_batch`).

**Comment by AffanBinFaisal** [comment]
Opened feature request: #24479 — optional \calibration_batch\ for INT8 export, scoped per @raimbekovm's note on static vs dynamic TensorRT.

**Comment by Y-T-G** [comment]
Hello.

I am not sure if you can run calibration at a different batch size than what the model can take in. If the model was exported with batch 4, it wouldn't be able to take in a batch of 8 or 2. Unless you're using dynamic batch size.

Also Ultralytics uses MINMAX_CALIBRATION strategy which is not affected by batch size and the recommendation about batch size is not applicable to it. It only applies to Int8EntropyCalibrator.

**Comment by kfrohwergmh** [comment]
Ah, I see. That means, conversely, that unlike during training mode, the batch size is fixed, if i dont want to allow dynamic batch size. 
Does the calibration data loader then use the model’s Predict/Eval mode, so to speak, to calculate the quantization,  feeding the images in one by one (with a batch size of 1)?

My original understanding was that the dataset is used to determine the maximum and minimum values, which are then taken as the boundaries for the quantization mapping (min/max clipped to -128, 127).

**Comment by glenn-jocher** [comment]
Yes, that’s essentially correct: INT8 calibration runs inference-style passes over representative validation images with no training/updates, and if you export with `batch=1` then the calibration loader also feeds 1 image per step because `get_int8_calibration_dataloader()` uses `self.args.batch` for calibration too; with Ultralytics’ TensorRT export this uses `MINMAX_CALIBRATION`, so the data is used to estimate activation ranges/scales rather than to change weights. See the [TensorRT INT8 export guide](https://docs.ultralytics.com/integrations/tensorrt) and the [`get_int8_calibration_dataloader()` reference](https://docs.ultralytics.com/reference/engine/exporter).

**Comment by kfrohwergmh** [comment]
Alright, thank you very much :)

**Comment by glenn-jocher** [comment]
You’re welcome — glad that clarified it.
